### ISSUE TITLE
Spurious error E0597 (does not live long enough)

### ISSUE THREAD

**Issue by Dridi** [rust-lang/rust#44019]
Please excuse the silly example, this is the minimal reproducer I came up with:

```rust
use std::rc::Rc;
use std::cell::RefCell;

type RcMyRef = Rc<RefCell<MyRef>>;

struct MyRef(isize);

impl MyRef {
    fn new_rc(&self) -> RcMyRef {
        Rc::new(RefCell::new(MyRef(self.0 + 1)))
    }
}

struct MyStruct(Option<RcMyRef>);

impl MyStruct {
    fn my_fn(&mut self) -> Option<RcMyRef> {
        let rc = self.0.take().unwrap();
        //return Some(rc.borrow().new_rc());
        Some(rc.borrow().new_rc())
    }
}

fn main() { }
```

If I try to build this code, I get this error:

```
$ rustc test.rs 
error[E0597]: `rc` does not live long enough
  --> test.rs:21:5
   |
20 |         Some(rc.borrow().new_rc())
   |              -- borrow occurs here
21 |     }
   |     ^ `rc` dropped here while still borrowed
   |
   = note: values in a scope are dropped in the opposite order they are created

error: aborting due to previous error(s)
```

However if I use the `return` syntax, it builds just fine (with lots of dead code warnings of course). I'm using Rust 1.19.0 packaged by Fedora.

**Comment by Mark-Simulacrum** [comment]
I believe this is "expected" behavior though isn't ideal from the users perspective. I think there's a more canonical issue, but can't find it right now.

**Comment by Dridi** [comment]
I found that puzzling at first, but I could make sense of it in a branch but not as the very last statement of a function. And even in the branch it was somewhat confusing because the overall `match` or `if` statement was the last (compound) statement in the function too.

It took me a fair share of head scratching to finally figure why a construct that was working in one place was rejected somewhere else :)

**Comment by kennytm** [comment]
@Mark-Simulacrum Do you know if NLL will fix this?

**Comment by Mark-Simulacrum** [comment]
Not sure.

**Comment by arielb1** [comment]
It won't. The `Ref` is destroyed after the parent `Rc`, so this compiling would be unsound.
